best_practices.md

```diff
@@ -0,0 +1,262 @@
+            case 'A': case 'E': case 'I': case 'O': case 'U':
+                ++count;
+                break;
+            default:
+                // do nothing
+                break;
+        }
+    }
+    return count;
+}
+
+int main() {
+    std::string myString = "Hello, world!";
+    const char* cString = "Example string";  // C-style string
+
+    // Using string_view on a std::string
+    std::cout << "Vowels in '" << myString << "': " << count_vowels(myString) << std::endl;
+    // Using string_view on a C-style string
+    std::cout << "Vowels in '" << cString << "': " << count_vowels(cString) << std::endl;
+
+    return 0;
+}
+```
+
+
+## 4. Avoid Dynamic Allocations in Frequently Called Functions
+
+### Practice
+Try to avoid dynamic allocations in functions that are called every time a user runs an operation. If the vector size is known, use `std::array` or `std::tuple`. `ttnn::small_vector`(not implemented yet) is a good replacement for `std::vector` in 99% of use cases.
+
+### Explanation
+Dynamic allocations can be costly in terms of performance, especially in functions that are called frequently. Using fixed-size containers like `std::array` or `std::tuple` can significantly reduce overhead. `ttnn::small_vector` provides a performance-optimized alternative to `std::vector` for most scenarios.
+
+### Motivation
+- **Performance**: Reduces the overhead associated with dynamic memory allocation.
+- **Predictability**: Improves predictability of memory usage and performance.
+- **Efficiency**: Enhances overall efficiency by leveraging fixed-size containers or optimized small vectors.
+
+### Example
+```
+// Allocating on heap
+std::vector<int> dynamicVector = {1, 2, 3, 4, 5};
+
+// No heap allocation
+std::array<int, 5> fixedArray = {1, 2, 3, 4, 5};
+```
+
+## 5. Avoid Using `.at()` for Vector Access
+
+### Practice
+Avoid calling the `.at()` function on vectors and other stl containers. Perform all necessary checks before using the vector and use `TT_ASSERT` or `TT_FATAL`. The `.at()` function throws exceptions without context, which can be frustrating for users. Overall, avoid STL calls that throw exceptions. All inputs should be validated on our side.
+
+### Explanation
+The `.at()` function performs bounds checking and throws an exception if the index is out of range. However, the exceptions lack detailed context, making them hard to debug. By using `TT_ASSERT`, `TT_FATAL` or `TT_THROW`, you can provide clearer error messages and avoid exceptions.
+
+### Motivation
+- **User Experience**: Provides more informative error messages, improving user experience.
+- **Debugging**: Easier to debug issues with clear assertions and fatal errors.
+- **Performance**: Avoids the overhead associated with exception handling.
+
+### Example
+```
+for (int i = 0; i < fp32_vec.size(); i++) {
+  fp32_vec.at(i) = rand_float() + offset; // Very bad, breaks a lot of optimiations. Additional check.
```

Comment:
Can we get a better example? we rarely use at to assign a value like this and I afraid that it does not land the point well.
I'd rather see an example where we just read with .at vs []

Reply:
It is example from our codebase. But Will add one more.